QAchat/py/QAchat.ipynb

```diff
@@ -0,0 +1,361 @@
+    "Now, we will define a general evaluation function to test different prompts and pipelines using BrainTrust. This makes it easy to iterate and improve our AI app."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Use some of BrainTrust autoevals scoring functions\n",
+    "factuality = Factuality()\n",
+    "leven_evaluator = LevenshteinScorer()\n",
+    "\n",
+    "# Define custom scoring functions\n",
+    "def uses_bad_words_scorer(output):\n",
+    "    if \"ai language model\" in output.lower():\n",
+    "        return 0 # Bad LLM :(\n",
+    "    if \"sorry\" in output.lower():\n",
+    "        return 0 # Bad LLM :(\n",
+    "    if \"user\" in output.lower():\n",
+    "        return 0 # Bad LLM :(\n",
+    "    else:\n",
+    "        return 1\n",
+    "\n",
+    "def signature_scorer(output):\n",
+    "    if \"-BT\" in output:\n",
+    "        return 1 # Good LLM :)\n",
+    "    else:\n",
+    "        return 0\n",
+    "    \n",
+    "async def score_output(output, expected, input):\n",
+    "    # Score factuality using BrainTrust's factuality scorer\n",
+    "    # Learn more here: https://github.com/braintrustdata/autoevals/blob/main/templates/factuality.ya\n",
+    "    factuality_score = factuality(output, expected, input=input)\n",
+    "\n",
+    "    # Score the output using BrainTrust's Levenshtein scorer\n",
+    "    leven_score = leven_evaluator(output, expected)\n",
+    "\n",
+    "    bad_word_score = uses_bad_words_scorer(output)\n",
+    "    signature_score = signature_scorer(output)\n",
+    "\n",
+    "    return {\n",
+    "        factuality_score.name: factuality_score.score,\n",
+    "        leven_score.name: leven_score.score,\n",
+    "        \"bad_words\": bad_word_score,\n",
+    "        \"signature\": signature_score\n",
+    "    }\n",
+    "\n",
+    "    \n",
+    "# This is an evaluation script that tests a prompt and pipeline combination and logs the results to BrainTrust.\n",
+    "async def run_evaluation(generation_fn, experiment_name):\n",
+    "    # Initialize a BrainTrust experiment\n",
+    "    experiment = braintrust.init(project=PROJECT_NAME, experiment=experiment_name, dataset=dataset)\n",
+    "    \n",
+    "    output_tasks = [asyncio.create_task(generation_fn(test_case['input'])) for test_case in dataset]\n",
+    "    outputs_pairs = [await task for task in output_tasks]\n",
+    "    outputs = [output for output, _ in outputs_pairs]\n",
+    "    outputs_meta = [prompt for _, prompt in outputs_pairs]\n",
+    "\n",
+    "    scores_tasks = [asyncio.create_task(score_output(output, test_case['output'], test_case['input'])) for output, test_case in zip(outputs, dataset)]\n",
+    "    scores = [await task for task in scores_tasks]\n",
+    "    print(\"Finished scoring outputs\")\n",
+    "\n",
+    "    for test_case, output, score, meta in zip(dataset, outputs, scores, outputs_meta):\n",
```

Comment:
this ends up taking like 4m on my machine, presumably because the sentencepiece stuff that Chroma does is slow (and synchronous)?

Maybe we should run it on fewer examples (like 10 or 20) by default? We could have a constant at the top that is the total number to run on?

Reply:
Google colab and my machine took less than 1 minute